File: Data Structures & Algorithms/decode-ways/submission-5.py
class Solution:
    def numDecodings(self, s: str) -> int:
        if s[0] == '0':
            return 0

        cache = defaultdict(int)

        def helper(i):
            if i > len(s):
                return 0
            if i == len(s):
                return 1
            if i in cache:
                return cache[i]
            if int(s[i]) != 0:
                if int(s[i: i + 2]) <= 26:  
                    cache[i] = helper(i + 1) + helper(i + 2)
                else:
                    cache[i] = helper(i + 1)
            return cache[i]

        return helper(0)

File: Data Structures & Algorithms/anagram-groups/submission-5.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        int n = strs.length;
        if(n == 0)
            return new ArrayList<>();
        Map<String, List<String>> mp = new HashMap<>();
        for(String str: strs){
            String key = helper(str);
            if(!mp.containsKey(key))
                mp.put(key, new ArrayList<>());
            mp.get(key).add(str);
        }
        List<List<String>> result = new ArrayList<>();
        for(Map.Entry<String, List<String>> entry: mp.entrySet()){
            result.add(entry.getValue());
        }
        return result;
    }

    public String helper(String s){
        int[] counts = new int[26];
        for(char c: s.toCharArray())
            counts[c - 'a']++;
        StringBuilder sb = new StringBuilder();
        for(int count: counts)
            sb.append(count).append('#');
        return sb.toString();
    }
}

/*
strs = [] -> []
strs.length() == 1 -> return [[strs]];

mp = string(count): list<strings>
count = 26 elements
*/